src/stashcache.py

```diff
@@ -234,6 +241,109 @@ def generate_public_cache_authfile(vo_data, resource_groups, fqdn=None, legacy=T
     return authfile
 
 
+def generate_cache_scitokens(vo_data: VOsData, resource_groups: List[ResourceGroup], fqdn: str, suppress_errors=True) -> str:
+    """
+    Generate the SciTokens needed by a StashCache cache server, given the fqdn
+    of the cache server.
+
+    The scitokens config for a StashCache namespace is in the VO YAML and looks like:
+
+        DataFederations:
+            StashCache:
+                Namespaces:
+                    /store:
+                        - SciTokens:
+                            Issuer: https://scitokens.org/cms
+                            Base Path: /
+                            Restricted Path: /store
+
+    "Restricted Path" is optional.
+    `fqdn` must belong to a registered cache resource.
+
+    If suppress_errors is True, returns an empty string on various error conditions (e.g. no fqdn,
+    no resource matching fqdn, resource does not contain a cache server, etc.).  Otherwise, raises
+    ValueError or DataError.
+
+    """
+    template = """\
+[Global]
+audience = {resource.name}, https://{fqdn}
+
+{issuer_blocks_str}
+"""
+
+    if not fqdn:
+        if suppress_errors:
+            return ""
+        else:
+            raise ValueError("fqdn: empty")
+
+    resource = _get_cache_resource(fqdn, resource_groups, suppress_errors)
+    if not resource:
+        return ""
+
+    issuer_blocks = []
+    for vo_name, vo_data in vo_data.vos.items():
+        stashcache_data = vo_data.get('DataFederations', {}).get('StashCache')
+        if not stashcache_data:
+            continue
+
+        namespaces = stashcache_data.get("Namespaces", {})
+        if not namespaces:
+            continue
+
+        has_non_public = False
+        for authz_list in namespaces.values():
+            for authz in authz_list:
+                if authz != "PUBLIC":
+                    has_non_public = True
+                    break
+        if not has_non_public:
+            continue
+
+        if not _cache_is_allowed(resource, vo_name, stashcache_data,
+                                 public=False, suppress_errors=suppress_errors):
+            continue
+
+        for dirname, authz_list in namespaces.items():
+            for authz in authz_list:
+                if not isinstance(authz, dict) or not isinstance(authz.get("SciTokens"), dict):
+                    continue
+                issuer_blocks.append(_get_scitokens_issuer_block(vo_name, authz['SciTokens'],
+                                                                 dirname, suppress_errors))
+
+    issuer_blocks_str = "\n".join(issuer_blocks)
+
+    return template.format(**locals()).rstrip() + "\n"
+
+
+def _get_scitokens_issuer_block(vo_name: str, scitokens: Dict, dirname: str, suppress_errors: bool) -> str:
+    template = """\
+[Issuer {dirname}]
```

Comment:
Although I admit the issuer names are mostly arbitrary, using the directory name seems a bit strange.  I normally think of the issuer name as a human readable form of the `issuer` URL -- however, since there might need to be multiple issuer blocks per issuer, the `dirname` might be the best option.

Nothing to change, just noting something weird.

Reply:
blockname perhaps?

Reply:
Nah, we don't need a change.  Just wanted to write down my thoughts for posterity.